### PR TITLE
fs-1407-send-incomplete-applications-on-closure

### PR DESCRIPTION
**Pull request by gio-karunakaran**

### CODE REVIEW

File: requirements.in [hunk 1]
```diff
@@ -9,6 +9,7 @@ slugify
 python-dateutil
 invoke
 colored
+click
```
Comment (gidsg) [inline requirements.in:12]: Looks like this is unused, can we revert all the requirements changes?

File: scripts/send_application_on_closure.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            application["account_email"] = account_id.email
+            all_applications.append({"application": application})
+        count = 1
+        if len(all_applications) > 0:
+            for application in all_applications:
+                email = {
+                    "email": application.get("account_email")
+                    for application in application.values()
+                }
+                current_app.logger.error(
```
Comment (gidsg) [inline scripts/send_application_on_closure.py:59]: Can we log `info` instead of `error`?

File: scripts/send_application_on_closure.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                # Add new template in Govuk-notify & in Notification service
+                Notification.send(
+                    template_type=Config.NOTIFY_TEMPLATE_SUBMIT_APPLICATION,
+                    to_email=email.get("email"),
+                    content=application,
+                )
+                count += 1
+
+        else:
+            return "There are no applications to be sent."
```
Comment (gidsg) [inline scripts/send_application_on_closure.py:73]: I think these strings will never be printed. Can we log instead?

File: requirements.in (hunk 1 above)
Comment (gio-karunakaran) [reply requirements.in]: have removed this now :)

File: scripts/send_application_on_closure.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+    current_date_time = (
+        datetime.now().replace(microsecond=0).strftime("%Y-%m-%d %H:%M:%S")
+    )
+
+    fund_rounds = helpers.get_data(
+        Config.FUND_STORE_API_HOST
+        + Config.FUND_ROUND_ENDPOINT.format(fund_id=fund_id, round_id=round_id)
+    )
+    if current_date_time > fund_rounds.get("deadline"):
```
[the remark this answers is not in the document]
Comment (gio-karunakaran) [reply scripts/send_application_on_closure.py:27]: we did this as sort of a security measure? we didn't want someone to accidentally execute the action before the deadline had actually past? can get rid of it if you think it's overkill?

File: scripts/send_application_on_closure.py (hunk 2 above)
Comment (gio-karunakaran) [reply scripts/send_application_on_closure.py]: good spot have changed this! 👍🏾

File: scripts/send_application_on_closure.py (hunk 3 above)
Comment (gio-karunakaran) [reply scripts/send_application_on_closure.py]: done :)

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -7,10 +7,10 @@ repos:
...
         - --line-length=79
         - --experimental-string-processing
--   repo: https://gitlab.com/pycqa/flake8
-    rev: 3.9.2
-    hooks:
-    - id: flake8
+# -   repo: https://gitlab.com/pycqa/flake8
+#     rev: 3.9.2
+#     hooks:
+#     - id: flake8
```
Comment (gidsg) [inline .pre-commit-config.yaml:13]: why was this commented out? Can we revert this?

File: requirements-dev.txt [hunk 6]
```diff
@@ -111,6 +111,10 @@ gitdb==4.0.9
     # via gitpython
 gitpython==3.1.27
     # via bandit
+greenlet==1.1.3
```
Comment (gidsg) [inline requirements-dev.txt:114]: The changes in `requirements.txt` and `requirements-dev.txt` should also be reverted.

File: .pre-commit-config.yaml (hunk 5 above)
Comment (gio-karunakaran) [reply .pre-commit-config.yaml]: done!

File: requirements-dev.txt (hunk 6 above)
Comment (gio-karunakaran) [reply requirements-dev.txt:123]: have reverted this too now :)

File: db/models/applications.py [hunk 7]
```diff
@@ -1,5 +1,6 @@
 import random
 import string
+from typing import ParamSpec
```
Comment (gidsg) [inline db/models/applications.py:3]: This looks unused to me? Did you run `pre-commit`?

File: requirements-dev.txt [hunk 8]
```diff
@@ -14,6 +14,8 @@ aniso8601==9.0.1
     #   flask-restx
 asserts==0.11.1
     # via -r requirements-dev.in
+atomicwrites==1.4.1
```
Comment (gidsg) [inline requirements-dev.txt:17]: Can we revert all changes to requirements files? I don't think this PR requires any?

File: scripts/send_application_on_closure.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+                email = {
+                    "email": application.get("account_email")
+                    for application in application.values()
+                }
+                current_app.logger.info(
+                    f"Sending application {count} of"
+                    f" {len(all_applications)} to {email.get('email')}"
+                )
+
+                # Add new template in Govuk-notify & in Notification service
```
Comment (gidsg) [inline scripts/send_application_on_closure.py:64]: Can we prefix this comment with `TODO:`?

File: db/models/applications.py (hunk 7 above)
Comment (gio-karunakaran) [reply db/models/applications.py:3]: It does look unused, not sure how pre-commit did not catch this but have removed now 👍🏾

File: scripts/send_application_on_closure.py (hunk 9 above)
Comment (gio-karunakaran) [reply scripts/send_application_on_closure.py]: done 👍🏾

File: requirements-dev.txt (hunk 8 above)
Comment (gio-karunakaran) [reply requirements-dev.txt]: hopefully that should all be reverted now?? 🤞🏾

File: scripts/send_application_on_closure.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+import api.routes.application.helpers as helpers
+from app import app
+from config import Config
+from db.models.applications import ApplicationsMethods
+from db.models.forms import FormsMethods
+from external_services.models.notification import Notification
+from flask import current_app
+
+
+def send_email_on_deadline_task(fund_id, round_id):
```
Comment (gidsg) [inline scripts/send_application_on_closure.py:17]: Can we call this `send_incomplete_applications_after_deadline` or something more descriptive?

Comment (gio-karunakaran) [reply scripts/send_application_on_closure.py]: ahh yes, have changed so it's more descriptive now :)

**Review by gidsg — APPROVED** [review]
LGTM 👍 📦

**Review by harryyo — APPROVED** [review]
Looks good

**Pull request merged**